snowpark_style_guide.md

```diff
@@ -0,0 +1,413 @@
+```
+
+Apart from Python primitive types, Snowflake types (`Variant`, `Geography`, `MapType`, etc.)
+are available in the module `snowflake.snowpark.types`.
+
+```python
+from snowflake.snowpark.functions import udf
+from snowflake.snowpark.types import Variant, Geography, MapType
+
+@udf
+def typed_add(x: Variant, y: Geography) -> MapType():
+    return {1:None}
+```
+
+# User Defined Table Functions (UDTFs)
+
+Snowflake supports SQL UDFs that return a set of rows, consisting of 0, 1, or multiple rows,
+each of which has 1 or more columns. Such UDFs are called [tabular UDFs, table UDFs, or,
+most frequently, UDTFs (user-defined table functions)][udtf-service-doc].
+
+UDTFs can be registered similarly to UDFs by using the `snowflake.snowpark.Session.udtf.register` method
+or the `snowflake.snowpark.functions.udtf` function.
+
+```python
+from snowflake.snowpark.functions import udtf
+from snowflake.snowpark.types import IntegerType, StructField, StructType
+
+class GeneratorUDTF:
+    def process(self, n):
+        for i in range(n):
+            yield (i, )
+
+# using snowflake.snowpark.functions.udtf
+udtf(
+    GeneratorUDTF,
+    output_schema=StructType([StructField("number", IntegerType())]),
+    input_types=[IntegerType()]
+)
+
+# using snowflake.snowpark.Session.udtf.register
+session.udtf.register(
+   GeneratorUDTF,
+    output_schema=StructType([StructField("number", IntegerType())]),
+    input_types=[IntegerType()]
+)
+```
+
+Type hints work for UDTFs as well:
+
+```python
+from snowflake.snowpark.functions import udtf
+from typing import Iterable, Tuple
+
+@udtf(output_schema=["number"])
+class TypedGeneratorUDTF:
+    def process(self, n: int) -> Iterable[Tuple[int]]:
+        for i in range(n):
+            yield (i, )
+```
+
+# Stored Procedures
+
+[Stored procedures][sproc-service-doc] allow you to write procedural code that executes SQL.
+In a stored procedure, you can use programmatic constructs to perform branching and looping.
+
+Stored procedures can be registered similarly to UDFs and UDTFs by using the `snowflake.snowpark.Session.sproc.register` method.
+
+```python
+from snowflake.snowpark.functions import sproc
+from snowflake.snowpark.types import IntegerType
+
+def add(session_, x, y):
+    return session_.sql(f"select {x} + {y}").collect()[0][0]
+
+# using snowflake.snowpark.functions.sproc
+sproc(
+    add,
+    return_type=IntegerType(),
+    input_types=[IntegerType()]
+)
+
+# using snowflake.snowpark.Session.udtf.register
+session.sproc.register(
+    add,
+    return_type=IntegerType(),
+    input_types=[IntegerType()]
+)
+```
+
+Type hints work for stored procedures as well:
+
+```python
+from snowflake.snowpark.functions import sproc
+
+@sproc
+def typed_add(session_, x: int, y: int) -> int:
+    return session_.sql(f"select {x} + {y}").collect()[0][0]
+```
+
+# Miscellaneous
+- Test your code thoroughly with unit tests and integration tests. There are numerous [test cases in Snowpark][snowpark-tests] to
```

Comment:
Optionally, remove the hyphen that makes this render as a bullet point. None of the other sections are bulleted, so it looks a bit inconsistent.